### DISCUSSION TITLE
Internal handling of IV in AES_GCM_256, for multiple encryption calls on same plaintext

### DISCUSSION THREAD

**Discussion by prd-ld** [openssl/openssl#23637] [Q&A]
I am trying to understand the working of OpenSSL's EVP_EncryptUpdate function, specifically why does it return different cipher-text for same plaintext, key and IV (initially provided), over multiple calls. As far as I understand, the IV should be changed after each call, and the function takes care of the IV internally. The Question is how exactly is the IV handled internally? Trying to fetch the IV returns the same IV as the one the context is initialised with.

```` c
#include <openssl/ssl.h>

void print_array(char *title, unsigned char *buff, int len)
{
    printf("%s:\n", title);
    for(int i = 0; i<len; ++i)
        printf("%02x ", buff[i]);
    printf("\n");
}

void handleErrors()
{
    printf("Error in encryption/decryption\n");
}

void encrypt_EVP_aes_256_gcm_init(EVP_CIPHER_CTX **ctx, unsigned char
*key, unsigned char *iv)
{
    if(!(*ctx = EVP_CIPHER_CTX_new()))
        handleErrors();
    if(1 != EVP_EncryptInit_ex(*ctx, EVP_aes_256_gcm(), NULL, key, iv))
        handleErrors();
}

void encrypt(EVP_CIPHER_CTX *ctx, unsigned char *plaintext, int
plaintext_len, unsigned char *ciphertext, int *ciphertext_len)
{
    int len;
    if(1 != EVP_EncryptUpdate(ctx, ciphertext, &len, plaintext, plaintext_len))
        handleErrors();
    *ciphertext_len = len;
}

void test(const unsigned char* key, const unsigned char* iv, const unsigned char *data, int iters)
{
    unsigned char enc_data[32];

    int32_t cipherLength;
    EVP_CIPHER_CTX *encrypt_ctx;
    encrypt_EVP_aes_256_gcm_init(&encrypt_ctx, key, iv);

    for(int i = 0; i<iters; ++i)
    {
        print_array("iv", EVP_CIPHER_CTX_iv(encrypt_ctx), 12);    
        
        encrypt(encrypt_ctx, data, 32, enc_data, &cipherLength);

        print_array("enc data", enc_data, cipherLength);
    }
}

int main()
{
    const unsigned char* key = "\x3c\x57\x5e\x25\x5f\x43\x41\x69\x3d\x5e\x48\x29\x72\x54\x27\x55\x3e\x29\x28\x65\x31\x34\x4a\x3e\x52\x2f\x7c\x6a\x7b\x25\x78\x52";
    const unsigned char* iv = "\x75\x71\x71\x55\x36\x33\x59\x52\x2c\x22\x74\x7d\x6f\x36\x6d\x2e";
    const unsigned char data[32] = "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA";

    test(key, iv, data, 3);

    return 0;
}
````

**Comment by kroeckx** [comment]
Please see https://crypto.stackexchange.com/questions/5807/aes-gcm-and-its-iv-nonce-value for how GCM works.

**Reply by prd-ld** [reply]
Hi @kroeckx , thanks for the quick reply!

From whatever I have read, it seems that there exist guidelines on using the nonce, but no particular specification to be complied as a part of the algorithm. What I want to understand is how exactly is the IV managed within the OpenSSL implementation of GCM, once I have provided the initial IV and key. What changes happen internally that yield a separate cipher-text every time?

**Reply by t8m** [reply]
The IV is split into nonce and counter. The nonce is what you obtain with the EVP_CIPHER_CTX_iv() call. You do not get the counter which is the value that is incremented.

**Reply by prd-ld** [reply]
Hi @t8m, thanks for the reply. That was helpful in clearing things. Currently I am thinking what happens to the internal variables across multiple calls to the EVP_EncryptUpdate function. The counter seems to be maintained across all calls, i.e. if I have an input of length 32 bytes, the counter becomes 1->2->3. Calling the function again on the same input should now have counter 3->4->5, right?

I am working on a code to understand the internal workings of AES-GCM as implemented by OpenSSL. I am able to match the ciphertexts for lengths that are multiples of 16 Bytes, but not otherwise. How are the variables affected in cases where length % 16 != 0?

As far as I understand, the Cipher Block ( E (K, Yi) ) that is XORed with plaintext blocks is only dependent on the Key and IV. Given the Key and Nonce are fixed for this example, the only variable the determines the block would be the counter. I am unable to understand the behavior of the code in terms of updating the counter or other variables when the length is not a multiple of 16 bytes. Is there any other catch I am overlooking?
